### PR TITLE
feat(ecs): add forceNewDeployment option for Fargate and EC2 services

### LINKED ISSUE #27762: aws-ecs: FargateServiceProps Missing Field `forceNewDeployment`?
### Describe the bug

Hello,

I just noticed that my CDK script has been doing every step of my NodeJS backend deployment except for restarting my ECS service upon ECR image update. There seems to be numerous threads on this specific issue, and the common answer is to use `forceNewDeployment`, which can be set from the CLI as well as within the Fargate service definition.

Currently, I am able to go into the AWS dashboard and manually check `forceNewDeployment` but that isn't the programmatic solution I am looking for. Also, once AWS spins up the new service and removes the old one, I have to check the option again for the next deployment.

The real underlying issue is that, for some reason, the [FargateServiceProps interface](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.FargateServiceProps.html) has no mention of this specific option. I am really not sure why, given that this issue of updating the ECS service after an ECR update is pretty wide-spread.

### Expected Behavior

I should be able to set the `forceNewDeployment` to true for the service when it is first built and thereafter. I really just want the ECS service to restart without downtime when the ECR image updates. If there is another way to do this that I have missed, please do let me know.

### Current Behavior

`forceNewDeployment` is not included as an option when creating an ECS service with CDK. Even checking it manually in the console does not stick, as the current service will spin down after the new one (that has the option set to False by default) is spun up.

### Reproduction Steps

Not certain if there are reproduction steps--it moreso a lack of an important field within the service definition.

### Possible Solution

Allow `forceNewDeployment` to be set to true for the creation of every service, forcing new ECS services to take the place of old ones when a new image is pushed to ECR.

### Additional Information/Context

_No response_

### CDK CLI Version

2.103.1

### Framework Version

_No response_

### Node.js Version

20.9.0

### OS

Ubuntu

### Language

TypeScript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by yasomaru**
### Issue # (if applicable)

Closes #27762 .

### Reason for this change

CloudFormation now supports the ForceNewDeployment property for ECS services
  (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-service-forcenewdeployment.html). This allows users to trigger a new
  deployment when updating a service, even if the service definition hasn't changed. This is particularly useful when

  - Updating tasks to use a newer Docker image with the same tag (e.g., my_image:latest)
  - Rolling Fargate tasks onto a newer platform version
  - Forcing a redeployment without changing service parameters

### Description of changes

Added L2 construct support for forceNewDeployment in ECS services:

  - Added ForceNewDeployment interface with enabled and nonce properties
  - Added forceNewDeployment option to BaseServiceOptions (available for both FargateService and Ec2Service)
  - enabled defaults to true when forceNewDeployment is specified
  - nonce is optional and can be used to trigger deployments by changing its value (e.g., timestamp, commit hash)

```ts
new ecs.FargateService(this, 'Service', {
    cluster,
    taskDefinition,
    forceNewDeployment: {
      enabled: true,
      nonce: Date.now().toString(),
    },
  });

```

### Describe any new or updated permissions being added

No new IAM permissions required.

### Description of how you validated changes

 - Unit tests: Added 6 test cases covering:
    - Basic forceNewDeployment with both enabled and nonce
    - forceNewDeployment with only enabled: true
    - forceNewDeployment with only enabled: false
    - forceNewDeployment with only nonce (defaults enabled to true)
    - Ec2Service support
    - No forceNewDeployment specified (undefined behavior)
  - Integration test: Created integ.fargate-service-force-new-deployment.ts that deploys a Fargate service with forceNewDeployment enabled and verifies the
  service reaches ACTIVE state
  - README: Added documentation with usage examples

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts [hunk 1]
```diff
@@ -832,6 +875,10 @@ export abstract class BaseService extends Resource
       serviceRegistries: Lazy.any({ produce: () => this.serviceRegistries }, { omitEmptyArray: true }),
       serviceConnectConfiguration: Lazy.any({ produce: () => this._serviceConnectConfig }, { omitEmptyArray: true }),
       volumeConfigurations: Lazy.any({ produce: () => this.renderVolumes() }, { omitEmptyArray: true }),
+      forceNewDeployment: props.forceNewDeployment ? {
+        enableForceNewDeployment: props.forceNewDeployment.enabled ?? true,
+        forceNewDeploymentNonce: props.forceNewDeployment.nonce,
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts]: We should validate that if `forceNewDeployment.nonce` is set, it should be 1-255 chars long [docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/TemplateReference/aws-properties-ecs-service-forcenewdeployment.html).

File: packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts [hunk 2]
```diff
@@ -496,6 +525,20 @@ export interface BaseServiceOptions {
    */
   readonly canaryConfiguration?: TrafficShiftConfig;
 
+  /**
+   * Whether to force a new deployment of the service.
+   *
+   * By default, deployments aren't forced. You can use this option to start
+   * a new deployment with no service definition changes. For example, you can
+   * update a service's tasks to use a newer Docker image with the same
+   * image/tag combination (my_image:latest) or to roll Fargate tasks onto
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts]: Backticks.

```suggestion
   * image/tag combination (`my_image:latest`) or to roll Fargate tasks onto
```

File: packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts [hunk 3]
```diff
@@ -90,6 +90,35 @@ export interface DeploymentCircuitBreaker {
...
+ * Configuration for forcing a new deployment of the service.
+ */
+export interface ForceNewDeployment {
+  /**
+   * Whether to force a new deployment of the service.
+   *
+   * By default, deployments aren't forced. You can use this option to start
+   * a new deployment with no service definition changes. For example, you can
+   * update a service's tasks to use a newer Docker image with the same
+   * image/tag combination (my_image:latest) or to roll Fargate tasks onto
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts]: Backticks

```suggestion
   * image/tag combination (`my_image:latest`) or to roll Fargate tasks onto
```

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Ty. Just a couple of nits imo.

File: packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts (hunk 1 above)
Comment (yasomaru) [reply packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts]: Thanks for the suggestion! I've added validation to ensure `nonce` is between
  1 and 255 characters when specified.

  Added unit tests for:
  - Empty string (should throw)
  - Exceeds 255 characters (should throw)
  - Exactly 255 characters (boundary case, should pass)

File: packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts (hunk 3 above)
Comment (yasomaru) [reply packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts]: Fixed, thanks!

**Review by lpizzinidev — APPROVED** [review]
Nice.

**Comment by rgoltz** [comment]
Please mind the merged and released PR https://github.com/aws/aws-cdk/pull/35726 which might cover the same topic.

File: packages/aws-cdk-lib/aws-ecs/test/base-service.test.ts [hunk 4]
```diff
@@ -935,3 +935,265 @@ describe('forceNewDeployment', () => {
...
+    const template = Template.fromStack(stack);
+    template.hasResourceProperties('AWS::ECS::Service', {
+      ForceNewDeployment: {
+        EnableForceNewDeployment: true,
+        ForceNewDeploymentNonce: 'deployment-2024-01-01',
+      },
+    });
+  });
+
+  test('should work with Ec2Service', () => {
```
Comment (abidhasan-aws) [inline packages/aws-cdk-lib/aws-ecs/test/base-service.test.ts]: The name of the test here is very generic. Let's specify what we are testing here.

File: packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts [hunk 5]
```diff
@@ -95,6 +95,38 @@ export interface DeploymentCircuitBreaker {
...
+  /**
+   * Whether to force a new deployment of the service.
+   *
+   * When set to `true`, Amazon ECS will start a new deployment even if there
+   * are no changes to the service configuration. When set to `false`, the
+   * `ForceNewDeployment` property is explicitly set with `EnableForceNewDeployment: false`.
+   *
+   * @default true
+   */
+  readonly enabled?: boolean;
```
Comment (abidhasan-aws) [inline packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts]: This property is not optional in the corresponding [CloudFormation resource](https://docs.aws.amazon.com/AWSCloudFormation/latest/TemplateReference/aws-properties-ecs-service-forcenewdeployment.html). We should keep it that way.

`forceNewDeployment` inside `BaseServiceOptions` is already optional. So if a customer does not pass any object, CloudFormation does not set `ForceNewDeployment` at all. But if a customer decides to set `forceNewDeployment`, they should be required to set the `enabled` property, no silent defaults.

File: packages/aws-cdk-lib/aws-ecs/README.md [hunk 6]
```diff
@@ -2480,6 +2480,24 @@ service.forceNewDeployment();
 service.forceNewDeployment('my-custom-nonce-v2');
 ```
 
+Alternatively, you can configure `forceNewDeployment` declaratively as a constructor option.
```
Comment (abidhasan-aws) [inline packages/aws-cdk-lib/aws-ecs/README.md:2527]: Let add the fact that this can be overridden by the `forceNewDeployment` function.

File: packages/aws-cdk-lib/aws-ecs/README.md [hunk 7]
```diff
@@ -2480,6 +2480,24 @@ service.forceNewDeployment();
...
+declare const cluster: ecs.Cluster;
+declare const taskDefinition: ecs.TaskDefinition;
+
+// Force a new deployment with a stable nonce
+const service = new ecs.FargateService(this, 'Service', {
+  cluster,
+  taskDefinition,
+  forceNewDeployment: {
+    enabled: true,
+    nonce: 'v1.2.3',
```
Comment (abidhasan-aws) [inline packages/aws-cdk-lib/aws-ecs/README.md]: We can consider using a time-based nonce here instead of a hardcoded value, it better demonstrates the "force on every deploy" use case:
maybe : ```
nonce: Date.now().toString()```

File: packages/@aws-cdk-testing/framework-integ/test/aws-ecs/test/integ.fargate-service-force-new-deployment.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    nonce: 'initial-deployment',
+  },
+});
+
+const integTest = new IntegTest(app, 'ForceNewDeploymentTest', {
+  testCases: [stack],
+});
+
+// Verify the service is created and in ACTIVE state
+integTest.assertions.awsApiCall('ECS', 'describeServices', {
```
Comment (abidhasan-aws) [inline packages/@aws-cdk-testing/framework-integ/test/aws-ecs/test/integ.fargate-service-force-new-deployment.ts:42]: we can also check if `forceNewDeployment` property has set with correct values in this `integTest.assertions.awsApiCall`.

**Review by abidhasan-aws — DISMISSED** [review]

**Comment by abidhasan-aws** [comment]
Hi @yasomaru,
Thanks for the contribution :)
Added a few comments

File: packages/aws-cdk-lib/aws-ecs/test/base-service.test.ts (hunk 4 above)
Comment (yasomaru) [reply packages/aws-cdk-lib/aws-ecs/test/base-service.test.ts]: Renamed it to should set forceNewDeployment on an Ec2Service with enabled and nonce so the intent is explicit.

File: packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts (hunk 5 above)
Comment (yasomaru) [reply packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts]: Agreed. Made enabled required (dropped the ?? true fallback and the @default true tag), so once forceNewDeployment is set, enabled must be explicit, no silent default. forceNewDeployment itself stays optional,so omitting it leaves ForceNewDeployment unset. Updated the tests accordingly.

File: packages/aws-cdk-lib/aws-ecs/README.md (hunk 6 above)
Comment (yasomaru) [reply packages/aws-cdk-lib/aws-ecs/README.md:2527]: Added a note after the constructor example that forceNewDeployment() takes precedence over the constructor option, and  that the nonce passed to the method (or the auto-generated one) overrides any value set via the property.

File: packages/aws-cdk-lib/aws-ecs/README.md (hunk 7 above)
Comment (yasomaru) [reply packages/aws-cdk-lib/aws-ecs/README.md]: Switched the example to nonce: Date.now().toString() and updated  the comment to reflect the force-on-every-deploy use case.

File: packages/@aws-cdk-testing/framework-integ/test/aws-ecs/test/integ.fargate-service-force-new-deployment.ts (hunk 8 above)
Comment (yasomaru) [reply packages/@aws-cdk-testing/framework-integ/test/aws-ecs/test/integ.fargate-service-force-new-deployment.ts:42]: @abidhasan-aws 
I looked into asserting the deployed forceNewDeployment values, including actually deploying a getTemplate-based assertion, and there is no working way to do it. ECS read APIs do not return them (EnableForceNewDeployment/ForceNewDeploymentNonce are the write-only forceNewDeployment request parameter on CreateService/UpdateService,never persisted; DescribeServices, DescribeServiceDeployments, and Fn::GetAtt all lack them). The only API whose response contains the values is CloudFormation getTemplate, but deploying that assertion fails with ROLLBACK_COMPLETE: Response object is too long since it returns the entire template, and it would only re-read the synthesized template the snapshot already pins. So I kept the describeServices ACTIVE check and rely on the unit tests plus the snapshot for property-value
 correctness. Does this approach work for you, or is there a specific check you'd like me to add instead?

**Comment by yasomaru** [comment]
@abidhasan-aws 
Addressed all comments and synced with main. Ready for another look, thanks!

File: packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts [hunk 9]
```diff
@@ -98,6 +98,36 @@ export interface DeploymentCircuitBreaker {
...
+/**
+ * Configuration for forcing a new deployment of the service.
+ */
+export interface ForceNewDeployment {
+  /**
+   * Whether to force a new deployment of the service.
+   *
+   * When set to `true`, Amazon ECS will start a new deployment even if there
+   * are no changes to the service configuration. When set to `false`, the
+   * `ForceNewDeployment` property is explicitly set with `EnableForceNewDeployment: false`.
```
Comment (abidhasan-aws) [inline packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts]: The JSDoc for `enabled` property should clarify that setting `enabled: true` alone won't force a new deployment on subsequent updates. CloudFormation only triggers a redeployment when it detects a change in the template. The actual signal is the `ForceNewDeploymentNonce` value changing between deployments.

Could you update the JSDoc for `enabled` to something like:

```ts
/**
 * Whether to enable the force-new-deployment mechanism for the service.
 *
 * Setting this to `true` enables the mechanism, but a new deployment is only
 * triggered when the `nonce` value changes between deployments. If `nonce` is
 * not provided or its value remains the same across deployments, no new
 * deployment will be forced.
 *
 * To ensure a new deployment on every `cdk deploy`, provide a `nonce` with a
 * unique, time-varying value such as a timestamp, random string, or sequence
 * number (e.g., `Date.now().toString()`).
 *
 * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/TemplateReference/aws-properties-ecs-service-forcenewdeployment.html
 */
```

This prevents users from assuming `enabled: true` alone will force a redeployment on every `cdk deploy`.

**Review by abidhasan-aws — DISMISSED** [review]
Hi @yasomaru , 
Just left one comment!

File: packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts (hunk 9 above)
Comment (yasomaru) [reply packages/aws-cdk-lib/aws-ecs/lib/base/base-service.ts]: Thanks. Updated the enabled JSDoc to make clear thatenabling the mechanism alone does not force a redeployment on every cdk deploy: CloudFormation only starts a new deployment when it detects a template change, and the signal is the nonce changing. 
Added guidance to pass a time-varying nonce (e.g. Date.now().toString()) to force on every deploy, kept the enabled: false behavior note, and added the @see link.

**Review by abidhasan-aws — DISMISSED** [review]

**Review by abidhasan-aws — APPROVED** [review]
